tools/test_marker/main.py

```diff
@@ -0,0 +1,208 @@
+        return False
+
+    if is_attr:
+        return node.attr in patch_conf
+
+    return "RUSTPYTHON" in ast.unparse(node)
+
+
+def iter_patches(tree: ast.Module, conf: Conf) -> "Iterator[tuple[int, str]]":
+    """
+    Get needed patches to apply for given ast tree based on the conf.
+
+    Parameters
+    ----------
+    tree : ast.Module
+        AST tree to iterate on.
+    conf : Conf
+        Dict of `{ClassName: {test_name: Patch}}`.
+
+    Yields
+    ------
+    lineno : int
+        Line number where to insert the patch.
+    patch : str
+        Raw python code to be inserted at `lineno`.
+    """
+    # Phase 1: Iterate and mark existing tests
+    for key, nodes in ast.iter_fields(tree):
+        if key != "body":
+            continue
+
+        for i, cls_node in enumerate(nodes):
+            if not isinstance(cls_node, ast.ClassDef):
+                continue
+
+            if not (cls_conf := conf.get(cls_node.name)):
+                continue
+
+            for fn_node in cls_node.body:
+                if not isinstance(fn_node, ast.FunctionDef):
+                    continue
+
+                if not (patch_conf := cls_conf.pop(fn_node.name, None)):
+                    continue
+
+                if any(
+                    is_patch_present(dec_node, patch_conf)
+                    for dec_node in fn_node.decorator_list
+                    if isinstance(dec_node, (ast.Attribute, ast.Call))
+                ):
+                    continue
+
+                lineno = min(
+                    (dec_node.lineno for dec_node in fn_node.decorator_list),
+                    default=fn_node.lineno,
+                )
+
+                indent = " " * fn_node.col_offset
+                patch = format_patch(patch_conf)
+                yield (lineno - 1, f"{indent}{patch}")
+
+    # Phase 2: Iterate and mark inhereted tests
+    for key, nodes in ast.iter_fields(tree):
+        if key != "body":
+            continue
+
+        for i, cls_node in enumerate(nodes):
+            if not isinstance(cls_node, ast.ClassDef):
+                continue
+
+            if not (cls_conf := conf.get(cls_node.name)):
+                continue
+
+            for fn_name, patch_conf in cls_conf.items():
+                patch = format_patch(patch_conf)
+                yield (
+                    cls_node.end_lineno,
+                    f"""
+{INDENT1}{patch}
+{INDENT1}def {fn_name}(self):
+{INDENT2}return super().{fn_name}()
+""".rstrip(),
+                )
+
+
+def apply_conf(contents: str, conf: dict) -> str:
+    """
+    Patch a given source code based on the conf.
+
+    Parameters
+    ----------
+    contents : str
+        Raw python source code.
+    conf : Conf
+        Dict of `{ClassName: {test_name: Patch}}`.
+
+    Returns
+    -------
+    str
+        Patched raw python code.
+    """
+    lines = contents.splitlines()
+    tree = ast.parse(contents)
+
+    # Going in reverse to not distrupt the lines offset
```

Comment:
:broom: In the future it would be helpful to have `logger.debug()` output with before-after (plus some context lines) of the altered code. But assuming the list mutations are safe... that would be out-of-scope for a POC of-course.

@coderabbitai, please analysis this function (e.g., around lines 172-197) looking for possible improvements that could ensure any worries about list modifications here are mitigated.